sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -354,6 +354,14 @@ object TypeCoercion {
         p.makeCopy(Array(left, Cast(right, TimestampType)))
 
       case p @ BinaryComparison(left, right)
+        if left.isInstanceOf[AttributeReference] && right.isInstanceOf[Literal] =>
+        p.makeCopy(Array(left, castExpr(right, left.dataType)))
```

Comment:
We need to cover all the same cases, but it seems this fix couldn't do, for example;
```
scala> spark.udf.register("testUdf", () => "85908509832958239058032")
scala> sql("select * from values (1) where testUdf() > 1").explain
== Physical Plan ==
*Filter (cast(UDF:testUdf() as int) > 1)
+- LocalTableScan [col1#104]
```